### PR TITLE
Comprehensive Complex Numerics Support

### PR DESCRIPTION
**Pull request by narendasan**
# Description

Adding comprehensive support for decomposing complex numerics and running computation through Torch-TensorRT 

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Review by narendasan — COMMENTED** [review]
@apbose let me know if I added redundant infrastructure anywhere, I kind of worked from scratch

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 1]
```diff
@@ -587,6 +587,41 @@ def _save_weight_mapping(self) -> None:
                 weight_refit_map[engine_weight_name].dtype,
             ]
 
+        # Stage 3: Slice matching for unmatched non-scalar CONSTANT weights.
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:607]: @cehongwang please review

File: py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py [hunk 2]
```diff
@@ -169,103 +331,1490 @@ def replace_input_node(self, input_node: Node) -> None:
...
+                    logger.debug("  pass-through  %s  (elementwise-safe)", node.name)
+                else:
+                    logger.warning(
+                        "Complex op '%s' has no explicit rewrite rule. "
+                        "It will be passed through as-is on the real [..., 2] layout, "
+                        "which may produce incorrect results or fail TRT compilation. "
+                        "Consider adding a rewrite in complex_graph_rewrite.py.",
+                        node.target,
                     )
-
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py:210]:
```
if handler is not None and handler(node):
    modified = True
elif node.target in _ELEMENTWISE_SAFE:
``` 
we should change to this, so that a false in handler(node) should correctly log the passthrough. eg: add.Tensor with add.Tensor(tensor, tensor) should go to elementwise_safe requiring no rewrite.

File: py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py [hunk 3]
```diff
@@ -169,103 +331,1490 @@ def replace_input_node(self, input_node: Node) -> None:
...
                 new_node = self.gm.graph.get_attr(new_attr_name)
-        else:
-            logger.debug(
-                f"Unsupported node type in replacement of input node: {input_node.op}"
-            )
+            # Set fake-tensor metadata on the new node so that _is_complex_layout_node
+            # can identify it as a complex-layout [..., 2] tensor later when
+            # processing ops that use this buffer.
+            if fake_mode is not None:
+                try:
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py]: dont we need a similar fake_mode=FakeTensorMode() in buffer case as we did in the placeholder case?

File: py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py [hunk 4]
```diff
@@ -64,15 +145,19 @@ def subgraph_from_anchor(self, anchor_node: Node) -> ComplexSubGraphInfo:
...
+        # processes producers before consumers.  The set has no stable order,
+        # which caused bugs when e.g. mul(sin, sin) was processed before sin
+        # was rewritten (sin still had complex dtype, so the mul pattern ran
+        # against the original complex node and produced wrong results).
+        node_order = {n: i for i, n in enumerate(anchor_node.graph.nodes)}
+        ordered_subgraph = sorted(subgraph_nodes, key=lambda n: node_order.get(n, 0))
+        return ComplexSubGraphInfo([anchor_node], ordered_subgraph, list(input_nodes))
 
     def find_complex_op_subgraphs(
         self, gm: GraphModule, anchor_target: str
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py]: I assume this is no longer used?

File: examples/dynamo/torch_export_3d_rope.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+                dtype=torch.float32,
+            ),
+        ]
+    else:
+        trt_inputs = list(inputs)
+
+    trt_model = torch_tensorrt.dynamo.compile(
+        ep,
+        inputs=trt_inputs,
+        enabled_precisions={torch.float32},
```
Comment (zewenli98) [inline examples/dynamo/torch_export_3d_rope.py]: need to set `use_explicit_typing=False` or remove `enabled_precisions`

File: py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py [hunk 6]
```diff
@@ -169,103 +331,1490 @@ def replace_input_node(self, input_node: Node) -> None:
...
-                "This complex subgraph inputnode type does not need to replaced"
+                "  unpack get_attr  %s%s  ->  %s%s",
+                input_node.target,
+                tuple(original_tensor.shape),
+                new_attr_name,
+                tuple(stacked_tensor.shape),
             )
+        else:
+            pass  # call_function inputs are rewritten in-place by the op handlers
         input_node.replace_all_uses_with(new_node)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py:317]: `new_node` will not exist if `input_node.op` is not placeholder or get_attr

File: py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py [hunk 7]
```diff
@@ -169,103 +331,1490 @@ def replace_input_node(self, input_node: Node) -> None:
...
+            re1, im1 = self._inline_select_re_im(b, z1_inp)
+            re2 = b(torch.ops.aten.select.int, z2_inp, -1, 0)
+            im2 = b(torch.ops.aten.select.int, z2_inp, -1, 1)
+            log_re, log_im = self._inline_complex_log(b, re1, im1)
+            mul_re, mul_im = self._inline_complex_mul(b, re2, im2, log_re, log_im)
+            exp_re, exp_im = self._inline_complex_exp(b, mul_re, mul_im)
+            out = self._inline_cat_re_im(b, exp_re, exp_im)
+            node.replace_all_uses_with(out)
+            self.gm.graph.erase_node(node)
+            return True
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py:1271]: are these rewriting ops going to be added only for complex values and keep real values untouched?

File: tests/py/dynamo/models/test_model_refit.py [hunk 8]
```diff
@@ -1107,3 +1127,220 @@ def forward(self, x):
...
+    inputs = [z]
+
+    exp_program1 = torch.export.export(model1, tuple(inputs))
+    exp_program2 = torch.export.export(model2, tuple(inputs))
+
+    trt_gm = torchtrt.dynamo.compile(
+        exp_program1,
+        tuple(inputs),
+        use_python_runtime=True,
+        enabled_precisions={torch.float},
```
Comment (zewenli98) [inline tests/py/dynamo/models/test_model_refit.py]: same here

File: tests/py/dynamo/models/test_model_refit.py [hunk 9]
```diff
@@ -1107,3 +1127,220 @@ def forward(self, x):
...
```
Comment (zewenli98) [inline tests/py/dynamo/models/test_model_refit.py]: same here

File: tests/py/dynamo/models/test_model_refit.py [hunk 10]
```diff
@@ -1107,3 +1127,220 @@ def forward(self, x):
...
```
Comment (zewenli98) [inline tests/py/dynamo/models/test_model_refit.py]: same here

**Review by zewenli98 — COMMENTED** [review]
Is the newly added test folder hlo automatically running on CI?
A lot of changes involve. need to be careful to merge

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 11]
```diff
@@ -44,6 +44,7 @@
     resource_partition,
 )
 from torch_tensorrt.dynamo.utils import (
+    COMPLEX_TO_REAL_DTYPE,
     deallocate_module,
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/_compiler.py:47]: I believe this is not used in this file?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 12]
```diff
@@ -801,6 +802,104 @@ def compile(
...
+        output_node = list(partitioned_module.graph.nodes)[-1]
+        outputs = list(output_node.args[0])
+        for idx in complex_output_indices:
+            if idx >= len(outputs):
+                continue
+            src = outputs[idx]
+            if not isinstance(src, torch.fx.Node):
+                continue
+            if src.op == "call_module" and "_run_on_acc" in str(src.target):
+                with partitioned_module.graph.inserting_before(output_node):
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/_compiler.py:919]: shouldn't we have this check for `run_on_gpu` too? since we detect the complex graph for the whole graph, so the reshaping and modification to add 2 in shape (new shape = [...,2] would occur on all, so we would need to insert `view_as_complex` in the end to them too>?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/_compiler.py:919]: for run as gpu, it should just be in complex right? only TRT boundaries have the issue where we have this alternative representation for complex

File: py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py (hunk 7 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/lowering/passes/complex_graph_rewrite.py:1271]: yes, part of what the pass does is it identifies the boundaries of complex computation and only applies on that subgraph

**Review by cehongwang — APPROVED** [review]
LGTM

File: py/torch_tensorrt/dynamo/_refit.py [hunk 13]
```diff
@@ -587,14 +599,33 @@ def refit_module_weights(
...
+        else:
+            outputs_match = check_module_output(
+                new_module=new_gm,
+                refitted_module=compiled_module,
+                arg_inputs=torch_inputs,
+                kwarg_inputs=torch_kwarg_inputs,
+            )
+        if outputs_match:
             logger.info("Refitting Succeed!")
-            new_gm.to(CPU_DEVICE)
```
Comment (cehongwang) [inline py/torch_tensorrt/dynamo/_refit.py:599]: Don't we want to move this back to CPU?
